File: MKDS Course Modifier/Utility of Time CSharp/displaylist/DisplayListReader.cs
using System;
using System.Windows.Forms;

using Microsoft.VisualBasic;

using UoT.util;

namespace UoT {
  public static class DisplayListReader {
    // TODO: Remove combo box input.
    public static int ReadInDL(
        DlManager dlManager,
        uint address,
        ComboBox dListSelection) {
      IoUtil.SplitAddress(address, out var bank, out var offset);
      var data = Asserts.Assert(RamBanks.GetBankByIndex(bank));
      try {
        if (offset < data.Count) {
          // TODO: This jumps into the lowest level DL, but the 0xDE command (DL)
          // actually allows returning back up and calling more DLs. So this seems
          // like it will sometimes overlook any DLs that follow.
          // This should just be deleted and replaced w/ emulating in
          // F3DEX2_Parser.
          if (data[(int) offset] == 0xDE) {
            while (data[(int) offset] == 0xDE) {
              offset = IoUtil.ReadUInt24(data, (uint)(offset + 5L));
            }
          }

          int index = dlManager.Count;
          var displayList = new N64DisplayList();
          dlManager.Add(displayList);
          uint EPLoc = offset;
          dListSelection.Items.Add((index + 1).ToString() +
                                   ". " +
                                   Conversion.Hex(offset));
          displayList.StartPos = new ZSegment();
          displayList.StartPos.Offset = offset;
          displayList.StartPos.Bank = data.Segment;
          displayList.Skip = false;

          PickerUtil.NextRgb(out var r, out var g, out var b);
          displayList.PickCol = new Color3UByte {r = r, g = g, b = b};

          do {
            var commands = displayList.Commands;
            Array.Resize(ref commands, displayList.CommandCount + 1);
            displayList.Commands = commands;
            displayList.Commands[displayList.CommandCount] =
                new DLCommand(data, EPLoc);
            if (data[(int) EPLoc] == (int) F3DZEX.ENDDL | EPLoc >= data.Count) {
              EPLoc = (uint) (EPLoc + 8L);
              break;
            }

            EPLoc = (uint) (EPLoc + 8L);
            displayList.CommandCount += 1;
          } while (true);
          return (int) EPLoc;
        }
      } catch (Exception ex) {
        Interaction.MsgBox("Error reading in display list: " + ex.Message,
                           MsgBoxStyle.Critical,
                           "Exception");
        return default;
      }

      return default;
    }
  }
}

File: FinModelUtility/Utility of Time/Display List/DisplayListReader.vb
Module DisplayListReader
  ' TODO: Remove combo box input.
  Public Function ReadInDL(dlManager As DlManager, address As UInteger, dListSelection As ComboBox) As Integer
    Dim bank As Byte
    Dim offset As UInteger
    IoUtil.SplitAddress(address, bank, offset)

    Dim data As IBank = RamBanks.GetBankByIndex(bank)

    Try
      If offset < data.Count Then
        ' TODO: This jumps into the lowest level DL, but the 0xDE command (DL)
        ' actually allows returning back up and calling more DLs. So this seems
        ' like it will sometimes overlook any DLs that follow.
        ' This should just be deleted and replaced w/ emulating in
        ' F3DEX2_Parser.
        If data(offset) = &HDE Then
          Do Until data(offset) <> &HDE
            offset = IoUtil.ReadUInt24(data, offset + 5)
          Loop
        End If

        Dim index As Integer = dlManager.Count

        Dim displayList As New N64DisplayList
        dlManager.Add(displayList)

        Dim EPLoc As UInteger = offset

        dListSelection.Items.Add((index + 1).ToString & ". " & Hex(offset))

        With displayList
          .StartPos = New ZSegment
          .StartPos.Offset = offset
          .StartPos.Bank = data.Segment
          .Skip = False

          .PickCol = New Color3UByte
          PickerUtil.NextRgb(.PickCol.r, .PickCol.g, .PickCol.b)

          Do
            ReDim Preserve .Commands(.CommandCount)
            .Commands(.CommandCount) = New DLCommand(data, EPLoc)

            If data(EPLoc) = F3DZEX.ENDDL Or EPLoc >= data.Count Then
              EPLoc += 8
              Exit Do
            End If

            EPLoc += 8
            .CommandCount += 1
          Loop
        End With

        Return EPLoc
      End If
    Catch ex As Exception
      MsgBox("Error reading in display list: " & ex.Message, MsgBoxStyle.Critical, "Exception")
      Exit Function
    End Try
  End Function
End Module
